### ISSUE TITLE
0.6.0 External Metrics Bug around namespace being appended. 

### ISSUE THREAD

**Issue by john-delivuk** [kubernetes-sigs/prometheus-adapter#282]
Hey guys, I haven't had a moment to dig through the source yet and ensure that this is the confirmed case, but it looks like we're affixing namespace to the start of the external metrics queries. 

I get no errors in prom adapter but in the hpa I'm seeing the following.

```Metrics:                                                                        ( current / target )
  "aws_sqs_approximate_number_of_messages_visible_average" (target value):      45 / 1
  "aws_sqs_approximate_number_of_messages_not_visible_average" (target value):  5 / 1
Min replicas:                                                                   1
Max replicas:                                                                   5
Deployment pods:                                                                5 current / 5 desired
Conditions:
  Type            Status  Reason                   Message
  ----            ------  ------                   -------
  AbleToScale     True    SucceededGetScale        the HPA controller was able to get the target's current scale
  ScalingActive   False   FailedGetExternalMetric  the HPA was unable to compute the replica count: unable to get external metric [namespace]/aws_sqs_approximate_number_of_messages_visible_average/&LabelSelector{MatchLabels:map[string]string{queue_name: [ourQueueName],},MatchExpressions:[],}: no metrics returned from external metrics API
  ScalingLimited  True    TooManyReplicas          the desired replica count is more than the maximum replica count
Events:
  Type     Reason                   Age                   From                       Message
  ----     ------                   ----                  ----                       -------
  Warning  FailedGetExternalMetric  90s (x5501 over 23d)  horizontal-pod-autoscaler  unable to get external metric [namespace]/aws_sqs_approximate_number_of_messages_visible_average/&LabelSelector{MatchLabels:map[string]string{queue_name: [ourQueueName],},MatchExpressions:[],}: no metrics returned from external metrics API
```

Tonight I'll have time to dig into source and look at the PR's in the release, but I wanted to share my findings for all.

**Comment by john-delivuk** [comment]
also we should add TLS Certs to the base image as well. I'll submit a PR shortly.

**Comment by lawliet89** [comment]
Same issue here. I have downgraded to 0.5.0 and the issue went away.

Stacktrace: 

```
E0415 04:37:39.823155       1 provider.go:47] unable to generate a query for the metric: unable to convert resource namespaces into label: no generic resource label form specified for this metric
I0415 04:37:39.823475       1 wrap.go:47] GET /apis/external.metrics.k8s.io/v1beta1/namespaces/default/traefik_backend_requests_per_second: (2.79ms) 500
goroutine 2620362 [running]:
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).recordStatus(0xc0007f0c40, 0x1f4)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:204 +0xc8
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc0007f0c40, 0x1f4)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:183 +0x35
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/filters.(*baseTimeoutWriter).WriteHeader(0xc0004d1960, 0x1f4)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:227 +0xb2
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/metrics.(*ResponseWriterDelegator).WriteHeader(0xc000852e40, 0x1f4)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:376 +0x45
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.httpResponseWriterWithInit.Write(0x0, 0x1836d78, 0x23, 0x1f4, 0x1aa9240, 0xc0000caab0, 0xc000806280, 0x99, 0x9f, 0x0, ...)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:51 +0x1b9
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf.(*Serializer).Encode(0xc0000cf9c0, 0x1a8dc80, 0xc0006e0f00, 0x1a6b480, 0xc000852fc0, 0x148bc9d, 0x6)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go:218 +0x444
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning.(*codec).Encode(0xc0006907e0, 0x1a8dc80, 0xc0006e0f00, 0x1a6b480, 0xc000852fc0, 0x0, 0x0)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:215 +0x323
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.SerializeObject(0x1836d78, 0x23, 0x7f6462c395c8, 0xc0006907e0, 0x1aa9240, 0xc0000caab0, 0xc000901200, 0x1f4, 0x1a8dc80, 0xc0006e0f00)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:103 +0x11b
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObjectNegotiated(0x1aaae80, 0xc000532ea0, 0x1aab0c0, 0x27bb728, 0x1825dc6, 0x17, 0x1813865, 0x7, 0x1aa9240, 0xc0000caab0, ...)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:137 +0x555
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.ErrorNegotiated(0x1a69440, 0xc0006e0e60, 0x1aaae80, 0xc000532ea0, 0x1825dc6, 0x17, 0x1813865, 0x7, 0x1aa9240, 0xc0000caab0, ...)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:156 +0x167
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*RequestScope).err(...)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:83
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers.ListResource.func1(0x1aa9240, 0xc0000caab0, 0xc000901200)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/handlers/get.go:270 +0x11e5
github.com/directxman12/k8s-prometheus-adapter/vendor/github.com/kubernetes-incubator/custom-metrics-apiserver/pkg/apiserver/installer.restfulListResource.func1(0xc000852db0, 0xc000253960)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/github.com/kubernetes-incubator/custom-metrics-apiserver/pkg/apiserver/installer/installer.go:288 +0x8f
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/metrics.InstrumentRouteFunc.func1(0xc000852db0, 0xc000253960)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:272 +0x254
github.com/directxman12/k8s-prometheus-adapter/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc00003aa20, 0x7f6462a263c8, 0xc0002162a0, 0xc000901200)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/github.com/emicklei/go-restful/container.go:282 +0xa2e
github.com/directxman12/k8s-prometheus-adapter/vendor/github.com/emicklei/go-restful.(*Container).Dispatch(...)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/github.com/emicklei/go-restful/container.go:199
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x1829fdf, 0x1a, 0xc00003aa20, 0xc000340230, 0x7f6462a263c8, 0xc0002162a0, 0xc000901200)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/handler.go:146 +0x4d3
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1(0x7f6462a263c8, 0xc0002162a0, 0xc000901200)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/filters/authorization.go:64 +0x4f8
net/http.HandlerFunc.ServeHTTP(0xc00029b500, 0x7f6462a263c8, 0xc0002162a0, 0xc000901200)
        /usr/local/go/src/net/http/server.go:2007 +0x44
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/filters.WithMaxInFlightLimit.func1(0x7f6462a263c8, 0xc0002162a0, 0xc000901200)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/filters/maxinflight.go:160 +0x5e3
net/http.HandlerFunc.ServeHTTP(0xc0003d68a0, 0x7f6462a263c8, 0xc0002162a0, 0xc000901200)
        /usr/local/go/src/net/http/server.go:2007 +0x44
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1(0x7f6462a263c8, 0xc0002162a0, 0xc000901200)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go:50 +0x1eff
net/http.HandlerFunc.ServeHTTP(0xc00029b540, 0x7f6462a263c8, 0xc0002162a0, 0xc000901200)
        /usr/local/go/src/net/http/server.go:2007 +0x44
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthentication.func1(0x7f6462a263c8, 0xc0002162a0, 0xc000901000)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:81 +0x554
net/http.HandlerFunc.ServeHTTP(0xc0001220a0, 0x7f6462a263c8, 0xc0002162a0, 0xc000901000)
        /usr/local/go/src/net/http/server.go:2007 +0x44
github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1(0xc000ca7ce0, 0xc0000e36a0, 0x1ab2000, 0xc0002162a0, 0xc000901000)
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:112 +0xd0
created by github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP
        /go/src/github.com/directxman12/k8s-prometheus-adapter/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:99 +0x1cb
logging error output: "k8s\x00\n\f\n\x02v1\x12\x06Status\x12\x80\x01\n\x06\n\x00\x12\x00\x1a\x00\x12\aFailure\x1a0Internal error occurred: unable to fetch metrics\"\rInternalError*)\n\x00\x12\x00\x1a\x00\"\x1d\n\x00\x12\x17unable to fetch metrics\x1a\x00(\x002\x000\xf4\x03\x1a\x0
0\"\x00"
```

External Metrics Configuration

```yaml
    externalRules:
    - metricsQuery: sum by(backend) (rate(<<.Series>>{<<.LabelMatchers>>}[1m]))
      name:
        as: traefik_backend_requests_per_second
      seriesQuery: traefik_backend_requests_total{code != "429", code!~"5(.+)"}
```

**Comment by vranystepan** [comment]
I have the same issue, it's most likely caused by this MR, right?
https://github.com/DirectXMan12/k8s-prometheus-adapter/pull/197/files

**Comment by mentos1386** [comment]
```
  Warning  FailedGetExternalMetric       78s (x71 over 49m)    horizontal-pod-autoscaler  unable to get external metric mab/app_record_kinesis_external_events_scale_up/nil: unable to fetch metrics from external metrics API: Internal error occurred: unable to fetch metrics
```
Same issue. Using `v0.6.0` and `v0.7.0` on AWS EKS `1.14`

**Comment by john-delivuk** [comment]
Sorry I forgot to follow up. I'm going to take a peak tonight and see what the fix looks like

**Comment by john-delivuk** [comment]
Ok I'm remembering this a bit now, something that is poor judgement of the kubes api spec is that by default it's going to send `default` namespace for all external metric requests. Right now this is getting added to your query which causes it to not be found. A quick work around I haven't tested, is to add an add label to the prom adapter query using `label_replace`.


Problem section
https://github.com/DirectXMan12/k8s-prometheus-adapter/blob/master/pkg/naming/metrics_query.go#L147-L164

This is a little delicate. I'm thinking about how to future proof as much as possible

**Comment by john-delivuk** [comment]
@s-urbaniak would you prefer we go back to ignoring the namespace label on default or patch this with config? I'm not sure if there has been changes or plans to for the external metrics api.

**Comment by lawliet89** [comment]
Bump

**Comment by dgrisonnet** [comment]
Fixed by #380
